File: src/main/java/me/miki/shindo/management/security/impl/ParticleSecurity.java
package me.miki.shindo.management.security.impl;

import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventReceivePacket;
import me.miki.shindo.management.security.SecurityFeature;
import net.minecraft.network.play.server.S2APacketParticles;

public class ParticleSecurity extends SecurityFeature {

    private int particles;

    @EventTarget
    public void onReceivePacket(EventReceivePacket event) {
        if (event.getPacket() instanceof S2APacketParticles) {
            S2APacketParticles particle = ((S2APacketParticles) event.getPacket());

            particles += particle.getParticleCount();
            particles -= 6;
            particles = Math.min(particles, 150);

            if (particles > 100 || particle.getParticleCount() < 1 || Math.abs(particle.getParticleCount()) > 20 || particle.getParticleSpeed() < 0 || particle.getParticleSpeed() > 1000) {
                event.setCancelled(true);
            }
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/security/impl/ParticleSecurity.kt
package me.miki.shindo.management.security.impl

import me.miki.shindo.management.event.EventTarget
import me.miki.shindo.management.event.impl.EventReceivePacket
import me.miki.shindo.management.security.SecurityFeature
import net.minecraft.network.play.server.S2APacketParticles

class ParticleSecurity : SecurityFeature() {
    private var particles: Int = 0

    @EventTarget
    fun onReceivePacket(event: EventReceivePacket) {
        val pkt = event.getPacket()
        if (pkt is S2APacketParticles) {
            particles += pkt.particleCount
            particles -= 6
            particles = minOf(particles, 150)
            if (particles > 100 ||
                pkt.particleCount < 1 ||
                kotlin.math.abs(pkt.particleCount) > 20 ||
                pkt.particleSpeed < 0f ||
                pkt.particleSpeed > 1000f
            ) {
                event.setCancelled(true)
            }
        }
    }
}
